File: Sources/NativeRPCServiceCenter.m
//
//  NativeRPCServiceCenter.m
//  NativeRPCServiceKit
//
//  Created by FeliksLv on 2024/11/9.
//

#import "NativeRPCServiceCenter.h"
#import "NativeRPCAppService.h"

@interface NativeRPCServiceCenter ()

@property (nonatomic, strong) NSMutableDictionary<NSString *, Class> *serviceMap;

@end

@implementation NativeRPCServiceCenter

+ (NativeRPCServiceCenter *)sharedCenter {
    static NativeRPCServiceCenter *instance = nil;
    static dispatch_once_t onceToken;

    dispatch_once(&onceToken, ^{
        instance = [[NativeRPCServiceCenter alloc] init];
    });
    return instance;
}

- (instancetype)init {
    self = [super init];

    if (self) {
        _serviceMap = [NSMutableDictionary dictionaryWithDictionary:@{
            @"app": [NativeRPCAppService class],
        }];
    }

    return self;
}

- (void)registerService:(Class)service {
    if ([service conformsToProtocol:@protocol(NativeRPCService)]) {
        NSString *serviceName = [service serviceName];

        if (serviceName) {
            [_serviceMap setObject:service forKey:serviceName];
        }
    }
}

- (Class)serviceWithServiceName:(NSString *)serviceName {
    return [_serviceMap objectForKey:serviceName];
}

@end

File: Sources/NativeRPCServiceKit/NativeRPCServiceCenter.swift
//
//  NativeRPCServiceCenter.swift
//  NativeRPCServiceKit
//
//  Created by FeliksLv on 2025/1/12.
//

import Foundation

public final class NativeRPCServiceCenter {
    public static let shared = NativeRPCServiceCenter()
    
    private var serviceMap: [String: NativeRPCService.Type] = [
        NativeRPCEventService.name: NativeRPCEventService.self
    ]
    
    private init() {}

    private func registerService<T: NativeRPCService>(_ service: T.Type) {
        serviceMap[T.name] = service
    }

    private func serviceType(named serviceName: String) -> NativeRPCService.Type? {
        return serviceMap[serviceName]
    }

    public static func registerService<T: NativeRPCService>(_ service: T.Type) {
        shared.registerService(service)
    }

    static func serviceType(named serviceName: String) -> NativeRPCService.Type? {
        return shared.serviceType(named: serviceName)
    }
}
